src/annotations/mod.rs

```diff
@@ -0,0 +1,481 @@
+                S: serde::ser::Serializer,
+            {
+                use serde::ser::SerializeSeq;
+                let mut seq = serializer.serialize_seq(Some(self.time.len()))?;
+                for e in &self.time {
+                    seq.serialize_element(e)?;
+                }
+                seq.end()
+            }
+        }
+
+        impl Serialize for TaskId {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                usize::from(*self).serialize(serializer)
+            }
+        }
+
+        fn record_event(event: AnnotationEvent) {
+            with_state(move |state| {
+                let task_id = state.last_task_id.expect("no last task ID");
+
+                let task_id_num = usize::from(task_id);
+                assert!(task_id_num < state.tasks.len());
+                state.tasks[task_id_num].first_step = state.tasks[task_id_num].first_step.min(state.events.len());
+                state.tasks[task_id_num].last_step = state.tasks[task_id_num].last_step.max(state.events.len());
+
+                use std::backtrace::{Backtrace, BacktraceStatus};
+                use once_cell::sync::Lazy;
+                use regex::Regex;
+
+                // Here is a fragment of a backtrace for reference:
+                // ```
+                // 2: core::panicking::assert_failed_inner
+                // 3: core::panicking::assert_failed
+                //           at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:364:5
+                // 4: shuttle_clients::tests::example_impl
+                //           at ./src/example.rs:15:5
+                // 5: core::ops::function::Fn::call
+                //           at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:79:5
+                // ```
+                // We want to capture the frames (numbered lines above) which
+                // refer to files local to the project being run, as well as
+                // the function name, and line/column info.
+                // TODO: for now, "local to the project" is detected based on
+                //       the path starting with `./src/`. Find an alternative
+                //       way to do this?
+                // The following regex matches frames with local paths. We rely
+                // on the string format of the backtrace because there is no
+                // better API. At the time of writing, even the unstable feature
+                // `backtrace_frames` does not provide a better interface.
+                // https://doc.rust-lang.org/std/backtrace/struct.BacktraceFrame.html
+                static RE: Lazy<Regex> =
+                    Lazy::new(|| Regex::new(r"([0-9]+): ([^\n]+)\n +at (\./src/[^:]+):([0-9]+):([0-9]+)\b").unwrap());
+
+                // Whether or not the following call actually captures a backtrace
```

Comment:
I'm finding the code below hard to follow. Would be nice with some more comments, better names, more spacing etc.